### PR TITLE
Don't pre-complete too many writes

### PR DESCRIPTION
**Pull request by halter73**
#356

### CODE REVIEW

File: test/Microsoft.AspNet.Server.KestrelTests/SocketOutputTests.cs [hunk 1]
```diff
@@ -192,6 +192,83 @@ public void WritesDontCompleteImmediatelyWhenTooManyBytesIncludingNonImmediateAr
             }
         }
 
+        [Fact]
+        public void WritesDontGetCompletedTooQuickly()
+        {
+            // This should match _maxBytesPreCompleted in SocketOutput
+            var maxBytesPreCompleted = 65536;
```
Comment (cesarbs) [inline test/Microsoft.AspNet.Server.KestrelTests/SocketOutputTests.cs]: nit: 64 * 1024 makes it more clear it's 64KB.

File: test/Microsoft.AspNet.Server.KestrelTests/SocketOutputTests.cs [hunk 2]
```diff
@@ -192,6 +192,83 @@ public void WritesDontCompleteImmediatelyWhenTooManyBytesIncludingNonImmediateAr
             }
         }
 
+        [Fact]
+        public void WritesDontGetCompletedTooQuickly()
```
Comment (cesarbs) [inline test/Microsoft.AspNet.Server.KestrelTests/SocketOutputTests.cs]: You'll probably have to skip this test on Mono (https://github.com/aspnet/KestrelHttpServer/pull/373).

File: test/Microsoft.AspNet.Server.KestrelTests/SocketOutputTests.cs [hunk 3]
```diff
@@ -192,6 +192,83 @@ public void WritesDontCompleteImmediatelyWhenTooManyBytesIncludingNonImmediateAr
...
+                socketOutput.WriteAsync(buffer).ContinueWith(onCompleted);
+                // Assert
+                // The first write should pre-complete since it is <= _maxBytesPreCompleted.
+                Assert.True(completedWh.Wait(1000));
+                Assert.True(onWriteWh.Wait(1000));
+                // Arrange
+                completedWh.Reset();
+                onWriteWh.Reset();
+
+                // Act
```
Comment (cesarbs) [inline test/Microsoft.AspNet.Server.KestrelTests/SocketOutputTests.cs]: Break this into multiple tests so you act, assert and arrange only once in each one.

File: test/Microsoft.AspNet.Server.KestrelTests/SocketOutputTests.cs (hunk 2 above)
Comment (halter73) [inline test/Microsoft.AspNet.Server.KestrelTests/SocketOutputTests.cs]: This uses `MockLibuv`, so it should be fine.

File: test/Microsoft.AspNet.Server.KestrelTests/SocketOutputTests.cs (hunk 3 above)
Comment (halter73) [inline test/Microsoft.AspNet.Server.KestrelTests/SocketOutputTests.cs]: I'm not going to split this into multiple tests, but these comments aren't as clear as the could be.

The first act/assert/arrange is really still setup with some sanity checking.

**Pull request merged**